### PR TITLE
Add ZipArchiveEntry.IsEncrypted for password-protection detection

### LINKED ISSUE #57197: Detect password protected file with ZipArchive
Currently if a file is not able to be decompressed, there is no information given in the  exception regarding if it was password protected file.

Expected outcome: Provide the reason for failure in extracting text for zip files, eg: A file was password protected etc

### LINKED ISSUE #68897: [API Proposal]: Add property to ZipArchiveEntry to indicate if the entry is encrypted
### Background and motivation

We have received requests to be able to detect if a ZipArchive contains password-protected entries. Without this information, the consumer of a zip file cannot determine if the contents can be extracted ahead of time. During extraction of the entire archive or a single entry, an `InvalidDataException` is thrown if an entry is encrypted, but that exception can be thrown in other circumstances as well.

By adding a property to indicate if an entry is encrypted, the entries could be inspected ahead of time and/or skipped during extraction.

See https://github.com/dotnet/runtime/issues/57197 for more background on the requests.

### API Proposal

```csharp
namespace System.IO.Compression;

public class ZipArchiveEntry
{
    public bool IsEncrypted { get; }
}
```

### API Usage

```csharp
private bool ArchiveContainsEncryptedEntries(ZipArchive archice)
{
    foreach (var entry in archive.Entries)
    {
        if (entry.IsEncrypted)
        {
            return true;
        }
    }

    return false;
}
```

### Exposing an Archive-level Property
We could consider additionally including an archive-level property to indicate if any of the entries are encrypted. However, populating this property would require iterating over the entries as is shown in the usage example above. Until such an API request is received, I recommend we limit the new API surface to the entry-level property.

### Exposing the Encryption Level
Zip archives contain "weak" and "strong" encryption. Weak encryption (via the ZipCrypto algorithm) is indicated on a file with a single bit set within the General Purpose Bit Flags ([bit 0](https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT#:~:text=Bit%200%3A%20If%20set%2C%20indicates%20that%20the%20file%20is%20encrypted.)). Strong encryption (e.g. AES-*) is indicated with [both bits 0 and 6](https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT#:~:text=Bit%206%3A%20Strong%20encryption) being set in the General Purpose Bit Flags.

An alternative design would use an enum to indicate the encryption level of an entry:

* None
* Weak
* Strong

Unless we intend to offer APIs in the future that would enable the extraction of encrypted entries, the weak vs. strong indicator is unnecessary. With that consideration, I recommend we use a simple boolean to indicate if the entry is encrypted, regardless of the encryption level or algorithm. And with the boolean approach, we only need to check General Purpose Bit 0, since that bit must be set for both weak and strong encryption.

### netstandard support via Exception.Data
Some of the requests for this functionality came with the desire for netstandard support. Without adding new API surface, a failed extraction could include data on the exception instance that indicates why the `InvalidDataException` was thrown. For instance, `Exception.Data` could include a key/value pair of "IsEncrypted" and the boolean indicating if the entry was encrypted. However, this would rely on exception handling to control flow, which is an anti-pattern.

### Risks

The Zip file format also includes the ability to encrypt some of the central directory metadata, including file names. This API design does not take that concept into account. Popular tools such as WinZip, 7-Zip, and WinRAR do not support this capability either though. More typically, a password is set for the zip such that it is applied to each of the entries individually, and the central directory metadata remains unencrypted.

Based on the specification for how central directory metadata encryption is implemented, the entry count values are to be obfuscated if the central directory is encrypted ([reference](https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT#:~:text=7.3.6%20When%20the%20Central%20Directory%20is%20compressed%20and/or%20encrypted)). That expectation would likely introduce other issues with the current ZipArchive implementation if we sought to support encryption of the central directory. However, the file entry header General Purpose Bit Flags would remain the same, with bit 0 still indicating that the entry is also encrypted; thus the proposed design would still apply.

### PR DESCRIPTION
**Pull request by jeffhandley**
Fixes #57197, closes #68897. Uses test assets added in dotnet/runtime-assets#244.

This enables detection of password-protected zip archive entries, so that they can be skipped during processing of entries. Consumption code will generally look like this:

```c#
try
{
    if (!entry.IsEncrypted)
    {
        using (var entryStream = entry.Open())
        {
            entryStream.CopyTo(documentStream);
        }
    }
    else
    {
        // App-specific logic for how to handle/log that the zip contained encrypted entries
    }
}
catch (InvalidDataException)
{
    // This will no longer occur for encrypted entries since this code is skipping them
}
```

### CODE REVIEW

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs
Comment (Jozkee) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs:59]: This is faster than using `Enum.HasFlag`. See [assembly size in sharplab.io](https://sharplab.io/#v2:C4LghgzgtgNAJiA1AHwAICYAMBYAUBgRjzwDcwAnAAjIBtKBeSgIQEtgAxGsAcwDUwaAVwCmEAHQBJCAFEAdgGNyATwAOwYXErJmbTj35DRYgCJhgYY6MUs1Ae3IBuPKgIBOABS0AlE9zF8bu4ARACyBCCUQZSIlGGeAl4+zoGh6BFRMSHo8TSJvngARra2dHGsHFx8AiIQlNzCssLkAgAKguQqthDC5XrcXgwAfHUNTa3tnd29lWIAEpB97tP61UZScoqq6nBJfrhFJbHZy1WGtfWNzTRtHV09upUD9MPuF2PXE3cnlABkOhUrM6SGQKZRqDQDACEjEw+VwAG0+hAALp4BqCKD/PoGGqUCKCCAAC3swEoAG9KOtQVsNAxKJgAB4EGCUUzmSwQax2KgwhkADhZAFVZCx5LY4MJ2CwaMIAHJgKDCACCsjgAGFbFBFbJSby+ZhMJQAL54IA===).
```suggestion
            _isEncrypted = (_generalPurposeBitFlag & BitFlagValues.IsEncrypted) != 0;
```

**Comment by Jozkee** [comment]
CI errors are due files from runtime-assets are not being found in most CI legs.
@carlossanlop @dotnet/area-infrastructure-libraries, do you know how long it takes for the files merged in https://github.com/dotnet/runtime-assets/pull/244 to be fully available?
```
System.IO.Compression.Tests.zip_ReadTests.IdentifyEncryptedEntries(zipFile: "encrypted_entries_weak.zip") [FAIL]
      System.IO.FileNotFoundException : Could not find file 'C:\h\w\A38508CC\w\B24309D1\e\ZipTestData\refzipfiles\encrypted_entries_weak.zip'.
      Stack Trace:
        /_/src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Windows.cs(110,0): at Microsoft.Win32.SafeHandles.SafeFileHandle.CreateFile(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options)
        /_/src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Windows.cs(59,0): at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize)
        /_/src/libraries/System.Private.CoreLib/src/System/IO/File.cs(289,0): at System.IO.File.ReadAllBytes(String path)
        /_/src/libraries/Common/tests/System/IO/Compression/StreamHelpers.cs(12,0): at StreamHelpers.CreateTempCopyStream(String path)
        /_/src/libraries/System.IO.Compression/tests/ZipArchive/zip_ReadTests.cs(234,0): at System.IO.Compression.Tests.zip_ReadTests.IdentifyEncryptedEntries(String zipFile)
        --- End of stack trace from previous location ---
    System.IO.Compression.Tests.zip_ReadTests.IdentifyEncryptedEntries(zipFile: "encrypted_entries_aes256.zip") [FAIL]
      System.IO.FileNotFoundException : Could not find file 'C:\h\w\A38508CC\w\B24309D1\e\ZipTestData\refzipfiles\encrypted_entries_aes256.zip'.
      Stack Trace:
        /_/src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Windows.cs(110,0): at Microsoft.Win32.SafeHandles.SafeFileHandle.CreateFile(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options)
        /_/src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Windows.cs(59,0): at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize)
        /_/src/libraries/System.Private.CoreLib/src/System/IO/File.cs(289,0): at System.IO.File.ReadAllBytes(String path)
        /_/src/libraries/Common/tests/System/IO/Compression/StreamHelpers.cs(12,0): at StreamHelpers.CreateTempCopyStream(String path)
        /_/src/libraries/System.IO.Compression/tests/ZipArchive/zip_ReadTests.cs(234,0): at System.IO.Compression.Tests.zip_ReadTests.IdentifyEncryptedEntries(String zipFile)
        --- End of stack trace from previous location ---
    System.IO.Compression.Tests.zip_ReadTests.IdentifyEncryptedEntries(zipFile: "encrypted_entries_mixed.zip") [FAIL]
      System.IO.FileNotFoundException : Could not find file 'C:\h\w\A38508CC\w\B24309D1\e\ZipTestData\refzipfiles\encrypted_entries_mixed.zip'.
      Stack Trace:
        /_/src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Windows.cs(110,0): at Microsoft.Win32.SafeHandles.SafeFileHandle.CreateFile(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options)
        /_/src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Windows.cs(59,0): at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String fullPath, FileMode mode, FileAccess access, FileShare share, FileOptions options, Int64 preallocationSize)
        /_/src/libraries/System.Private.CoreLib/src/System/IO/File.cs(289,0): at System.IO.File.ReadAllBytes(String path)
        /_/src/libraries/Common/tests/System/IO/Compression/StreamHelpers.cs(12,0): at StreamHelpers.CreateTempCopyStream(String path)
        /_/src/libraries/System.IO.Compression/tests/ZipArchive/zip_ReadTests.cs(234,0): at System.IO.Compression.Tests.zip_ReadTests.IdentifyEncryptedEntries(String zipFile)
```

**Comment by ViktorHofer** [comment]
> @carlossanlop https://github.com/orgs/dotnet/teams/area-infrastructure-libraries, do you know how long it takes for the files merged in https://github.com/dotnet/runtime-assets/pull/244 to be fully available?

A new official build got triggered after the PR got merged and the new package is already available. Once per day a maestro dependency update PR then gets opened in dotnet/runtime to update the runtime-assets dependencies. Be on the look out for "Update dependencies ..." PRs.

File: src/libraries/System.IO.Compression/ref/System.IO.Compression.cs
Comment (stephentoub) [inline src/libraries/System.IO.Compression/ref/System.IO.Compression.cs:113]: Is there anywhere we should be consuming this, like the various `ZipFile.Extract*` methods?

Comment (jeffhandley) [reply src/libraries/System.IO.Compression/ref/System.IO.Compression.cs:113]: I considered that in the API proposal in the [Additional APIs to Consider section](https://github.com/dotnet/runtime/issues/68897#user-content-zipfile-accelerators). Any change to the existing `ZipFile.Extract*` methods' behavior could be breaking though, so I held off on changing any of the existing methods.  I got confirmation from a partner team that this simple boolean will substantially improve their scenarios, so I figured it best to hold off on other changes/additions until we implement [Add password to ZipArchive · Issue #1545](https://github.com/dotnet/runtime/issues/1545).

With that said, if you have specific recommendations here it wouldn't necessarily be breaking, I'm open to ideas.

**Comment by jeffhandley** [comment]
@Jozkee Thanks for the suggestions, including that of generating the reference source instead of updating it by hand. I ran that task and it picked up some other inconsistencies in the System.IO.Compression APIs too, so I've included those changes here too.

/cc @carlossanlop

**Comment by danmoseley** [comment]
Just do I understand, when an entry is encrypted does one extract it normally, eg to a MemoryStrean that app can decrypt from - ie the archive simply is exposing a helpful flag ? Or is the archive itself in a special state, perhaps with even the file name encrypted for example.

**Comment by jeffhandley** [comment]
@danmoseley The archive is in the exact same state; the file names are unencrypted. For ZipCrypto (weak) encryption, one could grab the stream and presumably decrypt it with an implementation of that algorithm and the password known. For those ZipCrypto entries, our current extraction APIs will copy out the encrypted bytes. For AES-256 entries, we currently throw if you try to extract them.

The Zip file format does also include the ability to encrypt some of the central directory metadata, including file names. This API design does not take that concept into account. Popular tools such as WinZip, 7-Zip, and WinRAR do not support this capability either though. More typically, a password is set for the zip such that it is applied to each of the entries individually, and the central directory metadata remains unencrypted.

Based on the specification for how central directory metadata encryption is implemented, the entry count values are to be obfuscated if the central directory is encrypted ([reference](https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT#:~:text=7.3.6%20When%20the%20Central%20Directory%20is%20compressed%20and/or%20encrypted)). That expectation would likely introduce other issues with the current ZipArchive implementation if we sought to support encryption of the central directory. However, the file entry header General Purpose Bit Flag would remain the same, with bit 0 still indicating that the entry is also encrypted; thus, the implementation here would still apply.

**Comment by danmoseley** [comment]
> @danmoseley The archive is in the exact same state; the file names are unencrypted. For ZipCrypto (weak) encryption, one could grab the stream and presumably decrypt it with an implementation of that algorithm and the password known. For those ZipCrypto entries, our current extraction APIs will copy out the encrypted bytes. For AES-256 entries, we currently throw if you try to extract them.

Thanks -- I should have read the issue as you already explained it well there. Clearly this API addresses the original ask, and apparently the [internal requests](https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1486845) none of which are asking to extract the content.

> For AES-256 entries, we currently throw if you try to extract them.

Why the difference -- is it because in this case, the compression is before encryption? Just curious, clearly it doesn't matter today.

File: src/libraries/System.IO.Compression/ref/System.IO.Compression.cs
Comment (jeffhandley) [reply src/libraries/System.IO.Compression/ref/System.IO.Compression.cs]: Unrelated changes from generatereferenceassemblysource have been reverted

**Comment by jeffhandley** [comment]
@danmoseley With AES-256 encryption, the extraction needs to take a slightly different approach:

        Bit 6: Strong encryption.  If this bit is set, you MUST
               set the version needed to extract value to at least
               50 and you MUST also set bit 0.  If AES encryption
               is used, the version needed to extract value MUST 
               be at least 51. See the section describing the Strong
               Encryption Specification for details.  Refer to the 
               section in this document entitled "Incorporating PKWARE 
               Proprietary Technology into Your Product" for more 
               information.

With weak ZipCrytpo encryption, the compressed data is encrypted as well, but our underlying implementation is not checking that and it results in essentially extracting garbage data. One could argue that this inconsistent behavior is a bug and that attempting to extract weak encrypted entries should also throw. I expect we'll reevaluate those behaviors if/when #1545 is addressed.

File: src/libraries/System.IO.Compression/ref/System.IO.Compression.cs
Comment (carlossanlop) [reply src/libraries/System.IO.Compression/ref/System.IO.Compression.cs]: > Any change to the existing ZipFile.Extract* methods' behavior could be breaking though

Agreed, it would be breaking. I think we should open a new issue to discuss if we want to modify the logic of these extraction methods so that encrypted entries get skipped (since we don't yet have an API to provide a password).

Here's a summary of where the changes would be done:

The `ZipFile.Extract*` and `ZipFileExtensions.Extract*` methods all eventually [call this](https://source.dot.net/#System.IO.Compression.ZipFile/System/IO/Compression/ZipFileExtensions.ZipArchiveEntry.Extract.cs,87):

```cs
ExtractRelativeToDirectory(this ZipArchiveEntry source, string destinationDirectoryName, bool overwrite)
```

The method then does some validations, then [extracts the entry here](https://source.dot.net/#System.IO.Compression.ZipFile/System/IO/Compression/ZipFileExtensions.ZipArchiveEntry.Extract.cs,117), which is where we would check for encryption:

```cs
source.ExtractToFile(fileDestinationPath, overwrite: overwrite);
```

**Review by carlossanlop — APPROVED** [review]
LGTM with all the suggestions addressed.

For the static extraction methods, I think we should open a new issue as suggested [here](https://github.com/dotnet/runtime/pull/70036#discussion_r888549271), for further discussion.

**Review by carlossanlop — COMMENTED** [review]
Update: I just noticed there are related CI failures that need to be addressed.

File: src/libraries/System.IO.Compression/ref/System.IO.Compression.cs
Comment (Jozkee) [reply src/libraries/System.IO.Compression/ref/System.IO.Compression.cs]: Why do we have this type as public in src but not defined in ref?

Comment (carlossanlop) [reply src/libraries/System.IO.Compression/ref/System.IO.Compression.cs]: That is a good question. Regardless of the answer, I don't think it should be added in this PR.

**Comment by Jozkee** [comment]
> Update: I just noticed there are related CI failures that need to be addressed.

https://github.com/dotnet/runtime/pull/69709 contains the new test data, it needs to be merged for CI to pass.

**Review by jozkee — APPROVED** [review]
LGTM, we just need to wait for https://github.com/dotnet/runtime/pull/69709, and re-run CI before merge.

**Review by carlossanlop — APPROVED** [review]

**Pull request merged**
